### ISSUE TITLE
Will estimator be discarded?

### ISSUE THREAD

**Issue by chengmengli06** [tensorflow/estimator#4]
What is the purpose to move estimator out of the main tensorflow repo? 
Will it be discaded in the future? If so, Keras will be the only supported high level apis?

**Comment by chengmengli06** [comment]
Will distributed training still rely on estimator, or there are Keras based wrappers for distributed training? @tanzhenyu

**Comment by tanzhenyu** [comment]
@bhack I guess it will be distributed with TF packages throughout 2.0. But I'm not sure on this --  please wait on TF 2.0 documentation.

**Comment by SumNeuron** [comment]
@yhliang2018 

I would like to earnestly discuss your comment as someone who is new to TF from other ML libraries and someone who also just spent a bunch of time learning about TF Estimators (and is hence waaaay to concerned that they will be scrapped or "unloved" in the future).

> estimator will continue to exist and especially useful for canned models, but Keras will be the default high level api for many features.

From TensorFlow information (articles, docs, videos, etc) it was my impression that Estimator was the "highest" level api and the goal was to allow people to make models that were more 1. more readily shareable and 2. translatable across the many TF formats (and possibly directly into TensorFlow js). Now I was clearly wrong about both of those things. 

I do not see how Estimator's provide any more benefit for canned models, even in the future, if all the focus and emphasis would be on TensorFlow's subset of Keras, as why not just export those "canned" models from Keras?  Why have two things that almost, if not exactly, do the same thing at the same level of api?

Second, I honestly am confused about TensorFlow's direction, especially in relation to Keras. I get that Keras is a high level library built on top of TF, however tf.keras is a subset of Keras (at least last time I checked) and does not have all of Keras' features built in. 

In short, why have tf.keras? Why not just draw a line in the sand and say high level stuff is Keras, go there, and low level stuff is tf. 


There are ML libraries that have seamless integration from their lowest apis to highest points. How come TF seems so fragmented?

As a new comer I mean no disrespect, obviously TF is a very rich and powerful library. I just want to get my bearings, after being so wrong about Estimators...

**Comment by SumNeuron** [comment]
@tanzhenyu 

I appreciate your quick and detailed feedback as well. I would like to get my bearing and invest my time in learning the parts of the framework which make the most sense going forward. 

I can understand and appreciate the goal of making research work easier. 

If the direction is to make Keras production level ready (despite Estimator support in TF v2), that I am misguided to spend time with Estimators as by TF v3 they may very well be discarded.

Since Keras is built on top of TensroFlow, I am still confused about the whole "make TF like Keras" aspect. Why not make Keras even better and just let TF be TensorFlow and adjust TF to work directly with Keras objects (since Keras is built on top of TensorFlow)? Or is the idea that Keras was successful so now TF will slowly update and consume Keras and eventually discontinue Keras as a standalone library at some point?

Again, reiterating point 3,  "exist throughout TF 2.0" to me means that the longer I spend time with Estimator, the sooner it will become obsolete and the less features / support it will get as it is becoming obsolete. 

4. There was some missing aspects in regards to multi-label problems, but I opted for estimator at the time, because I thought Estimators would be the "unit of currency" so to say of TensorFlow...

This leaves me to my new main question. If TensorFlow wants to be more like Keras, why use tf.keras? Why not just Keras?  

I know there are a lot of low level features in tf that are useful (e.g. TF Records can accelerate training times just by themselves if you bother to use them, despite their clunky, asymmetric api), but I am struggling to get the "pitch" so to say of TF as a ML library.

I can understand it if what you are trying to do is some kind of new trick with the gradients, or if it is for educational purposes, but outside of that I do not really get it if that makes sense?

Right now, as a tf-inexperienced newcomer, I see lack of consistency, scarce documentation, contributors who have outright told me they do not wish to fill in that documentation because it is "low level" (despite the fact that "high-level" apis such as tf.keras, tf.data and tf.estimator sometimes directly require low level apis) and feel exhausted. 

I can not even train a distributed Estimator and then load it in a non-distributed setting (via Estimator API, although I could add a low level export in the model_fn) due to manual device placement.



I just want to be able to make and share reusable models. They might have some fancy custom normalizations, losses, architectures, or even gradients. What part of TensorFlow should I use, knowing that at the end I want to export it and let anyone else load it or serve it in production?


An aside, I have a custom library called Feature Input / Output that unifies TF Records' asymmetric api, who do I talk to about that?


Again, I sincerely thank you for your time and input.

**Comment by bhack** [comment]
Also there are more general problem about the relationship between Keras and tf.keras like in https://github.com/tensorflow/tensorflow/issues/23041. So it is really confusing to the users upstream/downstream relationship and who is driving the evolution with all the side effects (bug report, feature request, design discussion/RFC, etc..).

**Comment by wenmengzhou** [comment]
@tanzhenyu  As you said,  Keras will be the high-level api for tensorflow。 Although estimator will exists throughout TF2.0， but when the function of tf.keras get mature , will estimator be deprecated?

As we are doing a lot of work based on tf.estimator,  should we begin to transfer all the work uing tf,keras or you will provide a tool to help user transfer code based on tf.estimator to tf.keras?
